datafusion/optimizer/src/single_distinct_to_groupby.rs

```diff
@@ -131,177 +126,190 @@ fn contains_grouping_set(expr: &[Expr]) -> bool {
-                            }
-                        })
-                        .unzip();
-
-                    // and they can be referenced by the alias in the outer aggr plan
-                    let outer_group_exprs = out_group_expr_with_alias
-                        .iter()
-                        .map(|(out_group_expr, _)| out_group_expr.clone())
-                        .collect::<Vec<_>>();
-
-                    // replace the distinct arg with alias
-                    let mut index = 1;
-                    let mut group_fields_set = HashSet::new();
-                    let mut inner_aggr_exprs = vec![];
-                    let outer_aggr_exprs = aggr_expr
-                        .iter()
-                        .map(|aggr_expr| match aggr_expr {
-                            Expr::AggregateFunction(AggregateFunction {
-                                func_def: AggregateFunctionDefinition::BuiltIn(fun),
-                                args,
-                                distinct,
-                                ..
-                            }) => {
-                                // is_single_distinct_agg ensure args.len=1
-                                if *distinct
-                                    && group_fields_set.insert(args[0].display_name()?)
-                                {
-                                    inner_group_exprs.push(
-                                        args[0].clone().alias(SINGLE_DISTINCT_ALIAS),
-                                    );
+                                    col(alias_str),
+                                    Some(qualified_name(qualifier, field.name())),
+                                ),
+                            )
+                        }
+                    })
+                    .unzip();
+
+                // replace the distinct arg with alias
+                let mut index = 1;
+                let mut distinct_aggr_exprs = HashSet::new();
+                let mut inner_aggr_exprs = vec![];
+                let outer_aggr_exprs = aggr_expr
+                    .into_iter()
+                    .map(|aggr_expr| match aggr_expr {
+                        Expr::AggregateFunction(AggregateFunction {
+                            func_def: AggregateFunctionDefinition::BuiltIn(fun),
+                            mut args,
+                            distinct,
+                            ..
+                        }) => {
+                            if distinct {
+                                debug_assert_eq!(
+                                    args.len(),
+                                    1,
+                                    "DISTINCT aggregate should have exactly one argument"
+                                );
+                                let arg = args.swap_remove(0);
+
+                                let expr_id = distinct_aggr_exprs.hasher().hash_one(&arg);
+                                if distinct_aggr_exprs.insert(expr_id) {
```

Comment:
This code appears to deduplicate distinct aggregate expressions by comparing their hash values rather than by equality. While expressions that hash to the same value are very likely the same expression, this isn't guaranteed I don't think. 

I think for correctness the distinct_agg_exprs should hold the arguments by value -- e.g. `distinct_agg_exprs.insert(arg.clone())` or use the display name as in the previous version

I realize this is another clone, but I think it is required for correctness

In general finding some way to compute a digest / fingerprint of an expression that can be used for equality would be great, but maybe it should be done as a separate project.

Reply:
Keeping the correctness is more important here 👍 

Since `arg.display_name` is a smaller clone than `arg.clone`, so I updated the code to use the display name as in the previous version. Updated in https://github.com/apache/datafusion/pull/10527/commits/2ba102f8ce5401be2a8cd27716585ba0a71d4081